File: 项目文件/frontend/src/api/link-relations.ts
import { request } from '../utils/request';
import type {
  LinkRelationCreate,
  LinkRelationListResponse,
  LinkedEntity,
} from '../types/link-relation';
import type { UnifiedResponse } from '../types/user';

export interface LinkRelationListParams {
  source_type: string;
  source_id: string;
}

export interface LinkRelationDeleteParams {
  source_type: string;
  source_id: string;
  target_type: string;
  target_id: string;
}

/** 创建双向关联（同时插入 source→target 与 target→source） */
export async function createLink(
  source_type: string,
  source_id: string,
  target_type: string,
  target_id: string,
): Promise<UnifiedResponse<LinkRelationListResponse>> {
  const data: LinkRelationCreate = { source_type, source_id, target_type, target_id };
  return request<LinkRelationListResponse>('/link-relations/', { method: 'POST', body: data });
}

export async function listLinkRelations(
  params: LinkRelationListParams,
): Promise<UnifiedResponse<LinkRelationListResponse>> {
  const searchParams = new URLSearchParams();
  searchParams.set('source_type', params.source_type);
  searchParams.set('source_id', params.source_id);
  const query = searchParams.toString();
  return request<LinkRelationListResponse>(`/link-relations/${query ? `?${query}` : ''}`);
}

export async function getLinkedEntities(
  params: LinkRelationListParams,
): Promise<UnifiedResponse<LinkedEntity[]>> {
  const searchParams = new URLSearchParams();
  searchParams.set('source_type', params.source_type);
  searchParams.set('source_id', params.source_id);
  const query = searchParams.toString();
  return request<LinkedEntity[]>(`/link-relations/linked-entities${query ? `?${query}` : ''}`);
}

export async function deleteLink(params: LinkRelationDeleteParams): Promise<UnifiedResponse<null>> {
  const searchParams = new URLSearchParams();
  searchParams.set('source_type', params.source_type);
  searchParams.set('source_id', params.source_id);
  searchParams.set('target_type', params.target_type);
  searchParams.set('target_id', params.target_id);
  const query = searchParams.toString();
  return request<null>(`/link-relations/${query ? `?${query}` : ''}`, { method: 'DELETE' });
}

File: 项目文件/backend/app/api/link_relations.py
"""关联关系 API 路由"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user
from app.models.user import User
from app.schemas.common import UnifiedResponse
from app.schemas.link_relation import (
    LinkRelationCreate,
    LinkRelationListResponse,
    LinkRelationResponse,
)
from app.services.link_relation import (
    create_relation,
    delete_relation,
    get_linked_entities,
    list_relations,
)

router = APIRouter()


@router.post("/", response_model=UnifiedResponse[LinkRelationListResponse])
async def create_relation_endpoint(
    request: LinkRelationCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    relations = await create_relation(
        db,
        request.source_type,
        request.source_id,
        request.target_type,
        request.target_id,
    )
    return UnifiedResponse(
        data=LinkRelationListResponse(
            items=[LinkRelationResponse.model_validate(r) for r in relations],
            total=len(relations),
        )
    )


@router.get("/", response_model=UnifiedResponse[LinkRelationListResponse])
async def list_relations_endpoint(
    source_type: str = Query(...),
    source_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    relations, total = await list_relations(db, source_type, source_id)
    return UnifiedResponse(
        data=LinkRelationListResponse(
            items=[LinkRelationResponse.model_validate(r) for r in relations],
            total=total,
        )
    )


@router.get("/linked-entities", response_model=UnifiedResponse[list])
async def linked_entities_endpoint(
    source_type: str = Query(...),
    source_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    entities = await get_linked_entities(db, source_type, source_id)
    return UnifiedResponse(data=entities)


@router.delete("/", response_model=UnifiedResponse[None])
async def delete_relation_endpoint(
    source_type: str = Query(...),
    source_id: uuid.UUID = Query(...),
    target_type: str = Query(...),
    target_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    ok = await delete_relation(db, source_type, source_id, target_type, target_id)
    if not ok:
        raise HTTPException(status_code=404, detail="关联关系不存在")
    return UnifiedResponse(msg="关联关系已删除")
